source/GameGraphRoot.cpp

```diff
@@ -807,43 +751,40 @@ void GameGraphRoot::setSegHealthWarning(int alpha) {
 }
 
 void GameGraphRoot::doTeleportAnimation() {
-	if (ship->getStabilizerStatus() == ShipModel::StabilizerStatus::FAILURE &&
-		!prevIsStabilizerFail) {
-		// Start teleportation animation
+	StabilizerModel& stabilizer = ship->getStabilizer();
+
+	if (stabilizer.getState() == StabilizerModel::StabilizerState::Fail && !prevIsStabilizerFail) {
 		currentTeleportationFrame = 1;
-		ship->setStabilizerStatus(ShipModel::StabilizerStatus::ANIMATING);
 	}
+	prevIsStabilizerFail = stabilizer.getState() == StabilizerModel::StabilizerState::Fail;
+
 	if (currentTeleportationFrame != 0) {
-		// Continue teleportation animation
 		if (currentTeleportationFrame <= TELEPORT_FRAMECUTOFF_FIRST) {
-			stabilizerFailPanel->setVisible(true);
-			stabilizerFailText->setVisible(true);
+			// Nothing needs to happen here but this branch is left in to keep the logic of the next
+			// few more clear
 		} else if (currentTeleportationFrame <= TELEPORT_FRAMECUTOFF_SECOND) {
 			blackoutOverlay->setColor(Tween::fade(
 				Tween::linear(0, 1, currentTeleportationFrame - TELEPORT_FRAMECUTOFF_FIRST,
 							  TELEPORT_FRAMECUTOFF_SECOND - TELEPORT_FRAMECUTOFF_FIRST)));
 		} else {
 			if (currentTeleportationFrame == TELEPORT_FRAMECUTOFF_SECOND + 1) {
 				// Teleport models
-				for (const auto& donut : ship->getDonuts()) {
-					donut->teleport();
-				}
-				ship->setStabilizerStatus(ShipModel::StabilizerStatus::INACTIVE);
-				stabilizerFailPanel->setVisible(false);
-				stabilizerFailText->setVisible(false);
+				const auto& playerDonut =
+					ship->getDonuts().at(*MagicInternetBox::getInstance().getPlayerID());
+				playerDonut->teleport();
```

Comment:
I thought the point of the refactor was to fix our backwards dependencies. It would be nice if we could think of a way to not have the view directly modifying the models

Reply:
Yeah, I did think about this.
I didn't really prioritize refactoring the actual teleportation stuff for this PR; just getting the stabilizer display split out made this PR big enough. The other thing is that the way we have it right now actually is the easiest way to avoid either a circular dependency and / or duplicate code, as the simple alternative would require having the model and the view to both keep track of where they are in the animation.
The real solution is probably to pull the entire teleportation process out into a separate controller class, but I wasn't really focusing on the teleportation this PR. Maybe we should talk about it at the meeting on Sunday?

Reply:
yeah we can discuss it then

Reply:
I opened #404 so we don't forget.